pkg/controller/perconaservermongodb/users.go

```diff
@@ -0,0 +1,260 @@
+	internalSysSecretObj := corev1.Secret{}
+
+	err = r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      secretName,
+		},
+		&internalSysSecretObj,
+	)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return errors.Wrap(err, "get internal sys users secret")
+	}
+
+	if k8serrors.IsNotFound(err) {
+		internalSysUsersSecret := sysUsersSecretObj.DeepCopy()
+		internalSysUsersSecret.ObjectMeta = metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: cr.Namespace,
+		}
+		err = r.client.Create(context.TODO(), internalSysUsersSecret)
+		if err != nil {
+			return errors.Wrap(err, "create internal sys users secret")
+		}
+		return nil
+	}
+
+	// we do this check after work with secret objects because in case of upgrade cluster we need to be sure that internal secret exist
+	if cr.Status.State != api.AppStateReady {
+		return nil
+	}
+
+	newSysData, err := json.Marshal(sysUsersSecretObj.Data)
+	if err != nil {
+		return errors.Wrap(err, "marshal sys secret data")
+	}
+	newSecretDataHash := sha256Hash(newSysData)
+	dataChanged, err := sysUsersSecretDataChanged(newSecretDataHash, &internalSysSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "check sys users data changes")
+	}
+
+	if !dataChanged {
+		return nil
+	}
+
+	restartPods, err := r.manageSysUsers(cr, &sysUsersSecretObj, &internalSysSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "manage sys users")
+	}
+
+	internalSysSecretObj.Data = sysUsersSecretObj.Data
+	err = r.client.Update(context.TODO(), &internalSysSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "update internal sys users secret")
+	}
+
+	if restartPods {
+		err = r.restartPods(cr, newSecretDataHash)
+		if err != nil {
+			return errors.Wrap(err, "restart pods")
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaServerMongoDB) manageSysUsers(cr *api.PerconaServerMongoDB, sysUsersSecretObj, internalSysSecretObj *corev1.Secret) (bool, error) {
+	var (
+		sysUsers  []sysUser
+		userAdmin *sysUser
+	)
+	restartPods := false
+	for key := range sysUsersSecretObj.Data {
+		if bytes.Compare(sysUsersSecretObj.Data[key], internalSysSecretObj.Data[key]) == 0 {
+			continue
+		}
+		switch key {
+		case envMongoDBBackupPassword:
+			sysUsers = append(sysUsers, sysUser{
+				Name: string(sysUsersSecretObj.Data[envMongoDBBackupUser]),
+				Pass: string(sysUsersSecretObj.Data[envMongoDBBackupPassword]),
+			},
+			)
+			restartPods = true
+		case envMongoDBClusterAdminPassword:
+			sysUsers = append(sysUsers, sysUser{
+				Name: string(sysUsersSecretObj.Data[envMongoDBClusterAdminUser]),
+				Pass: string(sysUsersSecretObj.Data[envMongoDBClusterAdminPassword]),
+			},
+			)
+		case envMongoDBClusterMonitorPassword:
+			sysUsers = append(sysUsers, sysUser{
+				Name: string(sysUsersSecretObj.Data[envMongoDBClusterMonitorUser]),
+				Pass: string(sysUsersSecretObj.Data[envMongoDBClusterMonitorPassword]),
+			},
+			)
+		case envMongoDBUserAdminPassword:
+			userAdmin = &sysUser{
```

Comment:
why `userAdmin` var needed at all?
```suggestion
			sysUsers = append(sysUsers, &sysUser{
```

Reply:
here is a comment in the code https://github.com/percona/percona-server-mongodb-operator/blob/K8SPSMDB-233/pkg/controller/perconaservermongodb/users.go#L154